pkg/kubectl/cmd/util/editor/editor.go

```diff
@@ -0,0 +1,199 @@
+	}
+	// rather than parse the shell arguments ourselves, punt to the shell
+	shell := defaultEnvShell()
+	return append(shell, editor), true
+}
+
+func (e Editor) args(path string) []string {
+	args := make([]string, len(e.Args))
+	copy(args, e.Args)
+	if e.Shell {
+		last := args[len(args)-1]
+		args[len(args)-1] = fmt.Sprintf("%s %q", last, path)
+	} else {
+		args = append(args, path)
+	}
+	return args
+}
+
+// Launch opens the described or returns an error. The TTY will be protected, and
+// SIGQUIT, SIGTERM, and SIGINT will all be trapped.
+func (e Editor) Launch(path string) error {
+	if len(e.Args) == 0 {
+		return fmt.Errorf("no editor defined, can't open %s", path)
+	}
+	abs, err := filepath.Abs(path)
+	if err != nil {
+		return err
+	}
+	args := e.args(abs)
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = os.Stdin
+	glog.V(5).Infof("Opening file with editor %v", args)
+	if err := withSafeTTYAndInterrupts(cmd.Run); err != nil {
+		if err, ok := err.(*exec.Error); ok {
+			if err.Err == exec.ErrNotFound {
+				return fmt.Errorf("unable to launch the editor %q", strings.Join(e.Args, " "))
+			}
+		}
+		return fmt.Errorf("there was a problem with the editor %q", strings.Join(e.Args, " "))
+	}
+	return nil
+}
+
+// LaunchTempFile reads the provided stream into a temporary file in the given directory
+// and file prefix, and then invokes Launch with the path of that file. It will return
+// the contents of the file after launch, any errors that occur, and the path of the
+// temporary file so the caller can clean it up as needed.
+func (e Editor) LaunchTempFile(prefix, suffix string, r io.Reader) ([]byte, string, error) {
+	f, err := tempFile(prefix, suffix)
+	if err != nil {
+		return nil, "", err
+	}
+	defer f.Close()
+	path := f.Name()
+	if _, err := io.Copy(f, r); err != nil {
+		os.Remove(path)
+		return nil, path, err
+	}
+	if err := e.Launch(path); err != nil {
+		return nil, path, err
+	}
+	bytes, err := ioutil.ReadFile(path)
+	return bytes, path, err
+}
+
+// withSafeTTYAndInterrupts invokes the provided function after the terminal
+// state has been stored, and then on any error or termination attempts to
+// restore the terminal state to its prior behavior. It also eats signals
+// for the duration of the function.
+func withSafeTTYAndInterrupts(fn func() error) error {
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, childSignals...)
+	defer signal.Stop(ch)
+
+	inFd := os.Stdin.Fd()
+	if !term.IsTerminal(inFd) {
+		if f, err := os.Open("/dev/tty"); err == nil {
+			defer f.Close()
+			inFd = f.Fd()
+		}
+	}
+
+	if term.IsTerminal(inFd) {
+		state, err := term.SaveState(inFd)
+		if err != nil {
+			return err
+		}
+		go func() {
+			if _, ok := <-ch; !ok {
+				return
+			}
+			term.RestoreTerminal(inFd, state)
+		}()
+		defer term.RestoreTerminal(inFd, state)
+		return fn()
+	}
+	return fn()
+}
+
+func tempFile(prefix, suffix string) (f *os.File, err error) {
```

Comment:
TempFile is limited and doesn't allow you to select a suffix.  You want the
suffix to work on Windows.

On Tue, Sep 8, 2015 at 9:28 PM, Jeff Lowdermilk <notifications@github.com>
wrote:

> In pkg/kubectl/cmd/util/editor/editor.go
> <https://github.com/kubernetes/kubernetes/pull/13604#discussion_r38999530>
> :
>
> > +		if err != nil {
> > +			return err
> > +		}
> > +		go func() {
> > +			if _, ok := <-ch; !ok {
> > +				return
> > +			}
> > +			term.RestoreTerminal(inFd, state)
> > +		}()
> > +		defer term.RestoreTerminal(inFd, state)
> > +		return fn()
> > +	}
> > +	return fn()
> > +}
> > +
> > +func tempFile(prefix, suffix string) (f *os.File, err error) {
>
> Why not just use ioutil.Tempfile
> <https://golang.org/pkg/io/ioutil/#TempFile> instead of copying it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13604/files#r38999530>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift